Announce.php

```diff
@@ -103,4 +105,87 @@ function schema() {
 			array( 'UpdateFunction', 'delete_orphans_dismissals' ),
 		);
 	}
+
+	/**
+	 * Add the RESTful routes handled by this plugin.
+	 *
+	 * @param string $p_event_name The event name
+	 * @param array  $p_event_args The event arguments
+	 * @return void
+	 */
+	public function routes( $p_event_name, $p_event_args ) {
+		$t_app = $p_event_args['app'];
+		$t_plugin = $this;
+		$t_app->group(
+			plugin_route_group(),
+			function() use ( $t_app, $t_plugin ) {
+				$t_app->post( '/dismiss/{context_id}', [$t_plugin, 'route_dismiss'] );
+			}
+		);
+	}
+
+	/**
+	 * RESTful route for Announcement dismissal.
+	 *
+	 * Return status
+	 * - 200 announcement was successfully dismissed
+	 * - 400 invalid context or announcement not dismissable
+	 * - 500 dismissal failed (db insert/update error)
+	 *
+	 * @param Psr\Http\Message\ServerRequestInterface $request
+	 * @param Psr\Http\Message\ResponseInterface $response
+	 * @param array $args
+	 * @return Psr\Http\Message\ResponseInterface
+	 */
+	public function route_dismiss( $request, $response, $args) {
+		$t_timestamp = time();
+		$t_user_id = auth_get_current_user_id();
+
+		# Set the reference Context Id for dismissal
+		$t_context_id = (int)$args['context_id'];
+
+		# Make sure the message context actually exists
+		# and that it can actually be dismissed (i.e. marked as dismissable,
+		# or  a time-to-live has been set).
+		$t_context = AnnounceContext::load_by_id( $t_context_id );
+		if( !$t_context || !$t_context->dismissable && $t_context->ttl == 0 ) {
+			return $response->withStatus( HTTP_STATUS_BAD_REQUEST );
+		}
+
+		plugin_push_current( $this->basename );
+		$t_dismissed_table = plugin_table( 'dismissed' );
+
+		# check for existing dismissal
+		$t_query = "SELECT * FROM {$t_dismissed_table}
+			WHERE context_id=" . db_param() . "
+			AND user_id=" . db_param();
+		$t_result = db_query( $t_query,
+			array( $t_context_id, $t_user_id )
+		);
+
+		if( db_num_rows( $t_result ) < 1 ) {
+			$t_query = "INSERT INTO {$t_dismissed_table}
+				(context_id, user_id, timestamp)
+				VALUES
+				(" . db_param() . ", " . db_param() . ", " . db_param(
+				) . ")";
+			$t_param = array( $t_context_id, $t_user_id, $t_timestamp );
+		}
+		else {
+			$t_query = "UPDATE {$t_dismissed_table}
+				SET timestamp = " . db_param() . "
+				WHERE context_id=" . db_param() . " 
+				AND user_id=" . db_param();
+			$t_param = array( $t_timestamp, $t_context_id, $t_user_id );
+		}
+		if( db_query( $t_query, $t_param ) ) {
+			$t_status = HTTP_STATUS_SUCCESS;
+		} else {
+			$t_status = HTTP_STATUS_INTERNAL_SERVER_ERROR;
```

Comment:
It is generally helpful when you return failure status code like bad request, internal server error, etc to also return a string explaining the error.  So it is easy to understand what is going on from the client side.

Reply:
Thanks for the advice, will update the code accordingly.
Should the error message be localized, or is it OK to just use English ?

Reply:
English only - no localization for status code errors.